### PR TITLE
feat(totp): Add initial totp session verification logic

### LINKED ISSUE #301: TOTP Session Verification API
From breaking out https://github.com/mozilla/fxa-auth-db-mysql/issues/297, this issue tracks items needed to verify a session using TOTP. Once Auth-server has validated the totp code, this method is called to verify the session and set the verification method, etc.

### Alter Tables

Sessions Table

| Column        | Description           | Options  | Datatype|
| ------------- | ------------- | ----- | ------ |
| verificationMethod      | method used to verify session. ex, `email, email-2fa, totp-2fa` | NULL | BIGINT UNSIGNED|
| verifiedAt      | date session was verified      |   NULL | BIGINT UNSIGNED|
| mustVerify      | whether or not session needs to be verified before      |   NULL | BOOLEAN |

TOTP Table
| Column        | Description           | Options  | Datatype|
| ------------- | ------------- | ----- | ------ |
| verified      | token has been verified with a code | NULL | BOOLEAN |
| enable      |  TOTP is enabled/disabled | NULL | BOOLEAN |

#### .verifySessionWithMethod(uid, options)

Parameters:

* uid - (Buffer16) the uid of the account
* options - (Object) options
  * verificationMethod - `email, email-2fa, totp-2fa`
  * sessionTokenId - session 

Returns:

* resolves with:
    * an empty object `{}`
* rejects: with one of:
    * `error.notFound()`
    * `error.invalidCode()`
    * any error from the underlying storage engine

#### Verify Session from TOTP Code

* Method : `POST`
* Path : TBD
  * `uid` : hex128
  * `tokenVerificationId` : hex128

### PR DESCRIPTION
**Pull request by vbudhram**
This PR is the next stage for TOTP tokens. It adds the ability to verify TOTP tokens and sessions.

* Adds the `verified, enabled` property to TOTP tokens.
* Adds the `verificationMethod, verifiedAt` property to session tokens.
* Adds the `mustVerify` property to the session table. This property is in sync with the property on the unverified tokens table, ref: https://github.com/mozilla/fxa-auth-db-mysql/issues/301#issuecomment-366051763.
* Can be tested against https://github.com/mozilla/fxa-auth-server/pull/2321

Fixes #301

### CODE REVIEW

File: db-server/index.js [hunk 1]
```diff
@@ -125,6 +126,7 @@ function createServer(db) {
 
   api.get('/keyFetchToken/:id/verified', withIdAndBody(db.keyFetchTokenWithVerificationStatus))
   api.post('/tokens/:id/verify', withIdAndBody(db.verifyTokens))
+  api.post('/tokens/:id/verifyWith', withIdAndBody(db.verifyTokensWithMethod))
```
Comment (vbudhram) [inline db-server/index.js]: Updated this endpoint name to `/tokens/:id/verifyWith` and db to `verifyTokensWithMethod` over `verifySessionWithMethod`. This made sense because it verifies more than session tokens. Open to other names as well.

File: db-server/test/backend/remote.js [hunk 2]
```diff
@@ -1524,8 +1524,8 @@ module.exports = function(cfg, makeServer) {
             .spread((sessionTokenResp, keyFetchTokenResp) => {
               respOk(sessionTokenResp)
               respOk(keyFetchTokenResp)
-              const sessionToken = sessionTokenResp
-              const keyFetchToken = keyFetchTokenResp
+              const sessionToken = sessionTokenResp.obj
```
Comment (vbudhram) [inline db-server/test/backend/remote.js]: Unrelated, but noticed that this wasn't testing against the correct response object.

File: lib/db/mem.js [hunk 3]
```diff
@@ -38,7 +38,7 @@ var DEVICE_FIELDS = [
   'callbackIsExpired'
 ]
 
-var SESSION_FIELDS = [
+const SESSION_DEVICE_FIELDS = [
```
Comment (vbudhram) [inline lib/db/mem.js]: This is unrelated change that confused me while working on this PR. These are session device fields that are returned when updating/retrieving devices.

File: lib/db/schema/patch-072-073.sql [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    -- Verify tokens with tokenVerificationId
+    CALL verifyToken_3(@tokenVerificationId, @uid);
+
+    SET @updateCount = (SELECT ROW_COUNT());
+  COMMIT;
+
+  SELECT @updateCount;
+END;
+
+CREATE PROCEDURE `sessionWithDevice_12` (
```
Comment (vbudhram) [inline lib/db/schema/patch-072-073.sql]: The procedures below have been updated to return the `mustVerify` property of session tokens from either the session table or unverified tokens table.

**Comment by vbudhram** [comment]
@mozilla/fxa-devs r?

File: lib/db/mem.js [hunk 5]
```diff
@@ -549,13 +553,18 @@ module.exports = function (log, error) {
     item.accountCreatedAt = account.createdAt
 
     if (unverifiedTokens[id]) {
-      item.mustVerify = unverifiedTokens[id].mustVerify
+      // Coalesce `mustVerify`
+      if (sessionTokens[id].mustVerify) {
+        item.mustVerify = sessionTokens[id].mustVerify
+      } else {
+        item.mustVerify = unverifiedTokens[id].mustVerify
+      }
```
Comment (philbooth) [inline lib/db/mem.js]: This code confused me initially. Given that `item.mustVerify` is already initialised to `sessionTokens[id].mustVerify || null` on line 546, could it be simplified to a simple `if (! item.mustVerify) {` condition?

Comment (philbooth) [reply lib/db/mem.js]: s/simple/single/

File: db-server/test/backend/db_tests.js [hunk 6]
```diff
@@ -1761,7 +1761,7 @@ module.exports = function (config, DB) {
           })
           .then((session) => {
             // Returns verified session
-            assert.equal(session.mustVerify, null, 'mustVerify is not set')
+            assert.equal(!! session.mustVerify, false, 'mustVerify is false')
```
Comment (philbooth) [inline db-server/test/backend/db_tests.js]: Purely out of curiousity, do we actually need to keep using `!!` every time we test these properties?

`assert.equal` already uses `==` under the hood, so presumably `null` coerces to `false` anyway. And `true` is always `true`. Do any of them fail without the `!!`?

File: lib/db/schema/patch-072-073.sql [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+END;
+
+CREATE PROCEDURE `updateTotpToken_1` (
+  IN `uidArg` BINARY(16),
+  IN `verifiedArg` TINYINT(1),
+  IN `enableArg` TINYINT(1)
+)
+BEGIN
+
+    UPDATE `totp` SET verified = verifiedArg, enable = enableArg WHERE uid = uidArg;
```
Comment (philbooth) [inline lib/db/schema/patch-072-073.sql]: Pedantry: Looks like 2-space indentation elsewhere...

File: docs/DB_API.md [hunk 8]
```diff
@@ -838,3 +860,23 @@ Returns:
...
+
+  Update the TOTP token for the user.
+
+  Parameters:
+
+  * `uid` (Buffer16):
+    The uid of the owning account
+  * `tokenData`
+    * `verified`: Boolean whether TOTP token has been verified
+    * `enable`: Boolean whether TOTP token is enabled
```
Comment (philbooth) [inline docs/DB_API.md]: This only occurred to me while reviewing the auth server PR. Should it be `enabled` rather than `enable`? That way sounds more like a boolean state and less like an instruction to me.

File: lib/db/mem.js [hunk 9]
```diff
@@ -1207,12 +1216,31 @@ module.exports = function (log, error) {
...
+    const totpToken = totpTokens[uid]
+
+    if (! totpToken) {
+      return P.reject(error.notFound())
+    }
+
+    // Currently, users can only update the verified and enable flags.
+    // Updating shared secret and epoch will break clients.
+    totpToken.verified = token.verified
+    totpToken.enabled = token.enabled
```
Comment (philbooth) [inline lib/db/mem.js]: ...oh and here you are setting `enabled`, not `enable`. Is one of them wrong?

File: lib/db/schema/patch-072-073.sql [hunk 10]
```diff
@@ -0,0 +1,239 @@
+SET NAMES utf8mb4 COLLATE utf8mb4_bin;
+
+-- Add columns `verified` and `enable` to help manage TOTP token state
+ALTER TABLE `totp`
+ADD COLUMN `verified` TINYINT(1) DEFAULT FALSE,
+ADD COLUMN `enable` TINYINT(1) DEFAULT TRUE,
```
Comment (philbooth) [inline lib/db/schema/patch-072-073.sql]: ...yeah (he said, continuing the conversation with himself), I think `enabled` would be better actually, it seems more consistent with other column names.

File: docs/DB_API.md (hunk 8 above)
Comment (vbudhram) [reply docs/DB_API.md]: Yikes, looks I am not consistent with these fields/docs. I will update to use `enabled`.

File: db-server/test/backend/db_tests.js (hunk 6 above)
Comment (vbudhram) [reply db-server/test/backend/db_tests.js]: Unfortunately, this would break some tests because the mysql column is using `TINYINT(1)` instead of `BOOLEAN`. I don't believe there is any harm in using the boolean variable since it will return true/false instead of 0,1 which needed to be coerced. Will update!

Comment (vbudhram) [reply db-server/test/backend/db_tests.js]: > Unfortunately, this would break some tests because the mysql column is using TINYINT(1) instead of BOOLEAN

Turns that that was not correct and tests do pass. The only other explanation I have is that it was needed when we were using TAP. In either case, they have been updated.

**Comment by vbudhram** [comment]
@philbooth Thank you!

File: db-server/index.js [hunk 11]
```diff
@@ -97,6 +97,7 @@ function createServer(db) {
       return db.deleteEmail(req.params.id, req.params.email)
     })
   )
+
```
[the remark this answers is not in the document]
Comment (vbudhram) [reply db-server/index.js]: Interesting, looks like this might have been missed on the original secondary email implementation. Opened issue https://github.com/mozilla/fxa-auth-db-mysql/issues/311.

File: docs/API.md [hunk 12]
```diff
@@ -91,13 +91,15 @@ The following datatypes are used throughout this document:
     * forgotPasswordVerified    : `POST /passwordForgotToken/:id/verified`
 * Unverified tokens:
     * verifyTokens              : `POST /tokens/:tokenVerificationId/verify`
+    * verifyTokensWithMethod    : `POST /tokens/:tokenId/verifyWith`
```
[the remark this answers is not in the document]
Comment (vbudhram) [reply docs/API.md]: Yea this was intentional. The store procedure will look up the `tokenVerificationId` directly from the `tokenId`. The `tokenId` was needed because you need to set the verificationMethod on that session id as well.

[the remark this answers is not in the document]
Comment (vladikoff) [reply docs/API.md]: Cool 👍
